### PR TITLE
fix: DAG should not fail if its tasks have continueOn

### PR DESCRIPTION
**Pull request by simster7**
Fixes: #2653 

Do not consider a DAG as failed if all of its failed task have `continueOn` set for their phases.

### CODE REVIEW

File: workflow/controller/dag.go [hunk 1]
```diff
@@ -51,6 +51,15 @@ func (d *dagContext) taskNodeName(taskName string) string {
 	return fmt.Sprintf("%s.%s", d.boundaryName, taskName)
 }
 
+func (d *dagContext) getTaskFromNode(node *wfv1.NodeStatus) *wfv1.DAGTask {
```
Comment (alexec) [inline workflow/controller/dag.go:54]: tests?

Comment (simster7) [reply workflow/controller/dag.go:54]: Added

File: workflow/controller/dag.go [hunk 2]
```diff
@@ -51,6 +51,15 @@ func (d *dagContext) taskNodeName(taskName string) string {
...
+	for _, task := range d.tasks {
+		if node.Name == d.taskNodeName(task.Name) {
+			return &task
+		}
+	}
+	panic("task from node " + node.Name + " does not exist")
```
Comment (alexec) [inline workflow/controller/dag.go:60]: Panicking will crash the controller. Can this ever happen in the real world?

Comment (simster7) [reply workflow/controller/dag.go:60]: Nope, should be unreachable. A similar function is above this

**Comment by simster7** [comment]
Back-ported to 2.7

**Pull request merged**